embedded-graphics/src/primitives/rounded_rectangle.rs

```diff
@@ -0,0 +1,768 @@
+/// 50px height of the rectangle.
+///
+/// All corner radii will now be reduced by 2px (5 / 2, rounded down) resulting in the corner radii
+/// of (28px, 28px) for the top left corner and (23px, 23px) for all others.
+///
+/// ```rust
+/// use embedded_graphics::{
+///     prelude::*,
+///     primitives::{CornerRadii, Rectangle, RoundedRectangle},
+/// };
+///
+/// let radii = CornerRadii {
+///     top_left: Size::new(30, 30),
+///     top_right: Size::new(25, 25),
+///     bottom_right: Size::new(25, 25),
+///     bottom_left: Size::new(25, 25),
+/// };
+///
+/// let rectangle =
+///     RoundedRectangle::new(Rectangle::new(Point::new(5, 5), Size::new(50, 60)), radii);
+///
+/// assert_eq!(
+///     rectangle.corners,
+///     CornerRadii {
+///         top_left: Size::new(28, 28),
+///         top_right: Size::new(23, 23),
+///         bottom_right: Size::new(23, 23),
+///         bottom_left: Size::new(23, 23),
+///     }
+/// );
+///
+/// # Ok::<(), core::convert::Infallible>(())
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct RoundedRectangle {
+    /// The base rectangle
+    pub rectangle: Rectangle,
+
+    /// The radius of each corner
+    pub corners: CornerRadii,
+}
+
+impl Primitive for RoundedRectangle {
+    type PointsIter = Points;
+
+    fn points(&self) -> Self::PointsIter {
+        Points::new(self)
+    }
+}
+
+impl ContainsPoint for RoundedRectangle {
+    fn contains(&self, point: Point) -> bool {
+        if !self.rectangle.contains(point) {
+            return false;
+        }
+
+        let tl = self.get_corner_quadrant(Quadrant::TopLeft);
+
+        if tl.bounding_box().contains(point) {
+            return tl.contains(point);
+        }
+
+        let tr = self.get_corner_quadrant(Quadrant::TopRight);
+
+        if tr.bounding_box().contains(point) {
+            return tr.contains(point);
+        }
+
+        let br = self.get_corner_quadrant(Quadrant::BottomRight);
+
+        if br.bounding_box().contains(point) {
+            return br.contains(point);
+        }
+
+        let bl = self.get_corner_quadrant(Quadrant::BottomLeft);
+
+        if bl.bounding_box().contains(point) {
+            return bl.contains(point);
+        }
+
+        // We're in the rest of the rectangle at this point
+        true
+    }
+}
+
+impl Dimensions for RoundedRectangle {
+    fn bounding_box(&self) -> Rectangle {
+        self.rectangle
+    }
+}
+
+impl RoundedRectangle {
+    /// Creates a new rounded rectangle from a base rectangle and equal corner XY radius for all
+    /// corners.
+    pub fn with_equal_corners(rectangle: Rectangle, corner_radius: Size) -> Self {
+        Self::new(rectangle, CornerRadii::new_equal(corner_radius))
+    }
+
+    /// Creates a new rounded rectangle with different corner radii.
+    pub fn new(rectangle: Rectangle, corners: CornerRadii) -> Self {
+        Self {
+            rectangle,
+            corners: corners.confine(rectangle.size),
```

Comment:
Hm, they don't, and I'm not sure there's a good solution. Where would you suggest confining them? I could confine them in `Points::new()` or `StyledRoundedRectangleIterator::new()`, but that could allow a plain `RoundedRectangle` to be created in an invalid state. It would also require re-confining every time `.contains()` is called. 

What about adding another private struct, `ConfinedCornerRadii`? `CornerRadii::confine()` could produce a `ConfinedCornerRadii` which `RoundedRectangle::new()` accepts.

Reply:
The first question we need to answer is if it should be allowed to have overlapping radii in `RoundedRectangle`. One advantage of not confining the radii is that you can create a small rectangle with large radii and then scale it up by changing `rrect.rectangle.size` without loosing the original radii. This could, for example, be used in an animation of an expandable UI element.

> What about adding another private struct, `ConfinedCornerRadii`? `CornerRadii::confine()` could produce a `ConfinedCornerRadii` which `RoundedRectangle::new()` accepts.

I don't quite understand how this would work. Is the user supposed to call `confine` before passing the radii into `RoundedRectangle::new`?

Reply:
> The first question we need to answer is if it should be allowed to have overlapping radii in RoundedRectangle.

I don't think we should. It breaks the `ContainsPoint` results in unexpected ways. If we call `.confine()` in the `Points` and `StyledRoundedRectangle` iterator constructors, they would  produce a different shape than the theoretical representation in `RoundedRectangle`. To exemplify the `ContainsPoint` issue, `shape.contains()` would return a lot of false positives for the following rounded rect:

![image](https://user-images.githubusercontent.com/800791/81960057-49c87480-9608-11ea-8fa3-40dcbaf39e2a.png)

> This could, for example, be used in an animation of an expandable UI element.

In this specific example with the current behaviour, corner radii will scale with the shape as it animates, up to their configured size which would work well for a scaling animation.

> Is the user supposed to call confine before passing the radii into RoundedRectangle::new?

That was the idea, yeah. Not a great one so open to others :)

Reply:
> I don't think we should. It breaks the `ContainsPoint` results in unexpected ways. If we call `.confine()` in the `Points` and `StyledRoundedRectangle` iterator constructors, they would produce a different shape than the theoretical representation in `RoundedRectangle`. To exemplify the `ContainsPoint` issue, `shape.contains()` would return a lot of false positives for the following rounded rect:

I didn't mean that they should use different radii. All methods and iters, including `contains`, would also need to use the confined radii internally. Doing this for every point will be slow, but this could be improved later by adding the wrapper we discussed earlier.

> In this specific example with the current behaviour, corner radii will scale with the shape as it animates, up to their configured size which would work well for a scaling animation.

If you create a new `RoundedRectangle` this works, but not by reusing a rectangle and setting `rectangle.size` to a new value. Or am I missing something?

> That was the idea, yeah. Not a great one so open to others :)

IMO we would need to make the fields private, otherwise we can't guarantee that the radii are correctly confined.

Reply:
> If you create a new RoundedRectangle this works, but not by reusing a rectangle and setting rectangle.size to a new value.

Ah, I see what you mean now. Yes, that wouldn't work which is bad. I'll change the confine to the iters/contains code and just swallow the perf hit for now. This also means we can keep all the `RoundedRectangle` fields public.

Reply:
Changed. I removed the doc example as it's no longer possible to assert the confined corners, or show them in an example in a meaningful way.

The benchmark only slowed down by ~10% as well which is a pleasant surprise.

Reply:
> The benchmark only slowed down by ~10% as well which is a pleasant surprise.

Nice, that's far less than I expected.

I've opened #343 to discuss future performance improvements like the wrapper we discussed earlier.